src/main/kotlin/org/rust/lang/core/CFGBuilder.kt

```diff
@@ -0,0 +1,363 @@
+        loopScopes.push(loopScope)
+        callable()
+        loopScopes.pop()
+    }
+
+    private fun addAstNode(element: RsElement, vararg preds: CFGNode): CFGNode =
+        addNode(CFGNodeData.AST(element), *preds)
+
+    private fun addDummyNode(vararg preds: CFGNode): CFGNode =
+        addNode(CFGNodeData.Dummy, *preds)
+
+    private fun addUnreachableNode(): CFGNode =
+        addNode(CFGNodeData.Unreachable)
+
+    private fun addNode(data: CFGNodeData, vararg preds: CFGNode): CFGNode {
+        val newNode = graph.addNode(data)
+        preds.forEach { addContainedEdge(it, newNode) }
+        return newNode
+    }
+
+    fun addContainedEdge(source: CFGNode, target: CFGNode) {
+        val data = CFGEdgeData(emptyList())
+        graph.addEdge(source, target, data)
+    }
+
+    private fun addReturningEdge(fromNode: CFGNode) {
+        val data = CFGEdgeData(loopScopes.map { it.loop })
+        graph.addEdge(fromNode, exit, data)
+    }
+
+    private fun straightLine(expr: RsExpr, pred: CFGNode, subExprs: List<RsExpr?>): CFGNode {
+        val subExprsExit = subExprs.fold(pred) { acc, subExpr -> process(subExpr, acc) }
+        return addAstNode(expr, subExprsExit)
+    }
+
+    fun process(element: RsElement?, pred: CFGNode): CFGNode {
+        if (element == null) return pred
+
+        result = null
+        val oldPredsSize = preds.size
+        preds.push(pred)
+        element.accept(this)
+        preds.pop()
+        assert(preds.size == oldPredsSize)
+
+        return checkNotNull(result) { "Processing ended inconclusively" }
+    }
+
+    private fun allPats(pat: RsPat, subPats: List<RsPat?>): CFGNode {
+        val patsExit = subPats.fold(pred) { pred, subPat -> process(subPat, pred) }
+        return addAstNode(pat, patsExit)
+    }
+
+    private fun processCall(callExpr: RsExpr, funcOrReceiver: RsExpr?, args: List<RsExpr?>): CFGNode {
+        val funcOrReceiverExit = process(funcOrReceiver, pred)
+        return straightLine(callExpr, funcOrReceiverExit, args)
+    }
+
+    override fun visitBlock(block: RsBlock) {
+        val stmtsExit = block.stmtList.fold(pred) { pred, stmt -> process(stmt, pred) }
+        val blockExpr = block.expr
+        val exprExit = process(blockExpr, stmtsExit)
+
+        finishWithAstNode(block, exprExit)
+    }
+
+    override fun visitLetDecl(letDecl: RsLetDecl) {
+        val initExit = process(letDecl.expr, pred)
+        val exit = process(letDecl.pat, initExit)
+
+        finishWithAstNode(letDecl, exit)
+    }
+
+    override fun visitFieldDecl(fieldDecl: RsFieldDecl) = finishWith(pred)
+
+    override fun visitLabelDecl(labelDecl: RsLabelDecl) = finishWith(pred)
+
+    override fun visitExprStmt(exprStmt: RsExprStmt) {
+        val exit = process(exprStmt.expr, pred)
+        finishWithAstNode(exprStmt, exit)
+    }
+
+    override fun visitPatBinding(patBinding: RsPatBinding) =
+        finishWithAstNode(patBinding, pred)
```

Comment:
Looks like this method is never executed. We process RsPatBinding in `visitPatIdent` method (b/c RsPatBinding is always a child of RsPatIdent)

Reply:
Not really. It is used in the case of `RsPatField`. But it's enough weird. Maybe it's better to override `visitPatField`? and also please add a test for the "PatField" case.